### DISCUSSION TITLE
Lazer 2024.625.0 editor feedback

### DISCUSSION THREAD

**Discussion by RitaSummers** [ppy/osu#28669] [Ideas]
Hello, I'm an experienced mapper with a major focus on optimization and efficiency of workflow, and I have feedback regarding the lazer editor, primarily ergonomics. This list is vaguely sorted by importance.
	
	
	
Objects don't inherit the volume and sample set of the last objects. This is handled by the usage of green lines in stable, but in lazer, creating new objects results in objects with a Normal sample set at 100% volume every time you create every single object. I think the intended behavior is copying the settings of the last object, though this would also be not great, since this would discourage complex hit sounding systems that involve a lot of sample sets, volume changes and additions
Some of this is addressed by https://github.com/ppy/osu/discussions/28290, however hotkeys would speed up the process immensely, and their addition should not be ignored. Arguably, sectional settings are more comfortable, due to many maps modifying only a few objects in entire sections
![image](https://github.com/ppy/osu/assets/173868219/77a95221-c6ed-47d7-b7f6-a9bbfdc0cb24)
![image](https://github.com/ppy/osu/assets/173868219/7d0c56f5-9758-4239-83b1-ef97c67da8ba)

Timeline is too stacked with hit sound and slider velocity tags, so you can't immediately see ticks (you can deduce ticks by adjacent ticks, which is inconvenient). I think making ticks extend further from the timeline would be sufficient
![image](https://github.com/ppy/osu/assets/173868219/f723bdd4-0658-492e-8ad7-cd2350c28b80)

Anchor hit box is too small, considering the amount of times you would need to (quickly) hover over it.
Some of this is addressed by https://github.com/ppy/osu/issues/10991, however I think that the hit box of the anchor should be larger than it visually is (could say the same for objects on the playfield, but not as much)
![image](https://github.com/ppy/osu/assets/173868219/cf77f27e-6fa6-4d8a-a629-e17b3f54c06c)

Can't move pixel move objects with mouse due to strange snapping seemingly added to prevent accidental movement
https://github.com/ppy/osu/assets/173868219/6ba704c9-bede-4665-86d5-7d823ab23d39

Timeline hit sound tags are too small and inconvenient to use, because Ctrl/Alt + Q/W/E/R don't change sample set anymore. Selecting objects and slider segments on the playfield and changing the hit sound sample sets and additions with hotkeys is preferred, you can easily see which objects exactly you are changing. Adding slider segments to the top timeline would be helpful as well
Some of this is addressed by https://github.com/ppy/osu/discussions/20706
![image](https://github.com/ppy/osu/assets/173868219/fd6c0fa6-87ba-4d97-9a40-fe1615763051)
![image](https://github.com/ppy/osu/assets/173868219/48923552-37aa-421f-87af-ede267271495)

The bottom and top timelines overlay on top of objects, so you can't easily manipulate objects (slider anchors, primarily) off the grid that aren't truly offscreen (because the grid is not all of playable space, due to circle size and whatnot)
Some of this is addressed in https://github.com/ppy/osu/discussions/13189, however what is considered off-screen and off grid is very different and needs to be accounted for by developers
![image](https://github.com/ppy/osu/assets/173868219/98e09e3e-30bd-4668-bcf9-78937a5a438d)
![image](https://github.com/ppy/osu/assets/173868219/c21afeb6-6454-4cc9-83a8-a0a457147816)

The top and bottom bars are too thick compared to the space the text takes up, you could halve them in height and make more space for everything else
![image](https://github.com/ppy/osu/assets/173868219/8b41a098-2bf8-4e2e-a989-7682329f5637)
![image](https://github.com/ppy/osu/assets/173868219/e54f9aff-57f6-4b49-b5d3-613f72dfc927)

Afterimages last for less amount of time than approach circles (same amount on stable). Seems to be intentional on slider ends? I (and many other mappers) find slider end after images useful for many things, like judging structure, aligning patterns, etc.
![image](https://github.com/ppy/osu/assets/173868219/0a2de26a-b7a8-41f6-8686-4eabb112ba86)

You can't adjust distance snap or beat snap divisor with scroll wheel (less important to adjust beat snap divisor with scroll wheel because shift+1/2/3/4... works). Ctrl + mouse wheel changes volume instead of snap divisor, on stable volume only changes when cursor is off the grid. I don't see a reason to make this different from stable, I think it's important to mirror the workflow from stable, as that is where mappers have been developing muscle memory for the last 15 years

Sliders should remember the last anchor type that you used, or use a perfect curve by default in a 2 point slider, since adding another anchor to the slider will result in a red anchor 3 point, which is unlike the expected stable behavior.

RE: Ctrl+click to change anchor type - while it is cool that you can select multiple anchors at once, you can have both Ctrl click to change anchor type and select multiple anchors. To move multiple anchors at once, you need to select two or more anchors with Ctrl held down, if Ctrl is released at any point, only one anchor ends up being selected. This means that you can have the ability to cycle anchors by changing the anchor type the moment Ctrl is released after clicking, which is the expected stable behavior

Inspector only shows distance snap relative to previous object (and not also next like in stable)
![image](https://github.com/ppy/osu/assets/173868219/9ad43f8d-a38a-4980-9f6e-ae7275c6aeaa)

The break UI ruins the current usage of breaks as song representation because it's too busy. The break UI showing up is completely unnecessary on short breaks, and the behavior should be mirrored from stable, where short breaks only show the warning arrows, and not the section status (pass or fail), which would also include the remaining break time and whatnot on lazer. Arguably (and personally), break UI diminishes from the experience and song representation of certain long breaks, too. I think it ruins the anticipation and tension of the last break on [this map](https://osu.ppy.sh/beatmapsets/1726464#osu/3649088) (04:15:986), because the UI shows a lot of elements, and there is a lot more to look at, which may distract players from sitting down and focusing on listening to the song (because there's nothing else to look at, except maybe the background)
https://attendant.s-ul.eu/c4z2QDAJ.mp4
![image](https://github.com/ppy/osu/assets/173868219/e48ca4ee-bc92-4a07-a38f-856bdbc2934b)
![image](https://github.com/ppy/osu/assets/173868219/8f3c6c84-d01d-4213-b81b-561c818ff2c6)

Show green lines on the bottom timeline from legacy maps (and bookmark functionality is needed, but I think you're already working on that). Current lazer timelines are very empty, and it's subjectively difficult to navigate maps: personally, I don't have much problem with it, but several of my friends who watched me map on lazer couldn't comprehend where anything is in the map

I don't see a reason to have the select timeline to drag when that makes the hitbox for selecting multiple objects on the timeline smaller and harder to quickly select (this feature also is outclassed by simply scrolling with the mouse like we have for the last 15 years). I think it would be very frustrating if people wanted to quickly select multiple things on the timeline and accidentally dragged the timeline away from where they wanted to be, if you missed the timeline before dragging on stable, nothing would happen
Some of this is addressed by https://github.com/ppy/osu/issues/23784, however the dragging the timeline by holding the left mouse button seems absolutely pointless, because the wheel accomplishes the exact same thing in a much more precise way by snapping to ticks
(for some reason, this video refused to embed here) https://attendant.s-ul.eu/SNltf8ky.mp4

Holding modifiers (such as Ctrl, Alt, anything) on the timing screen to increase the bpm steps (0.1, 0.2, 0.5, 1 → 2, 3, 4, 5)
![image](https://github.com/ppy/osu/assets/173868219/9d772e82-54fe-45cc-92d4-830e3e8b6b09)

In stable, it is possible to click the time code in the bottom left to copy the time. This is mainly useful for modding and discussing maps with other people, because you can paste an object time code to instantly go to it with the pattern selected (00:27:815 (1,1,1,1) -)
![image](https://github.com/ppy/osu/assets/173868219/dbaa3772-ec99-48a1-ad22-da0655867820)
![image](https://github.com/ppy/osu/assets/173868219/bc03fc23-8b09-4d92-a776-ec315ee5538a)

Custom tick rates per slider (perhaps an object context menu option?)

No video support in the editor

These points are fully addressed by other issues:

~~Can't access the files in the map folder (you can't add samples, change background quickly with drag and drop, or access .osu)~~ https://github.com/ppy/osu/issues/12029 https://github.com/ppy/osu/issues/18902

~~Can't manipulate volumes mid-object (can't make slidertracks/spinners go louder or quieter as the object plays out)~~ https://github.com/ppy/osu/issues/9940

~~You can't set tick rate to 0.5 in the editor (due to .osu being inaccessible...)~~ https://github.com/ppy/osu/issues/28519

**Comment by peppy** [comment]
Thanks for the feedback.

On reading your feedback, much of what you have asked for is already tracked by issues. If you have the time, then please go through and remove any duplicated content and instead link to the existing issue to save us having to do this. If you don't have time, you can leave it how it is and we might get to it eventually.

**Comment by felipemarins** [comment]
List of duplicate issues:

> Can't access the files in the map folder

#12029 #18902

> Objects don't inherit the volume and sample set of the last objects.

#28290

> Anchor hit box is too small

I didn't find any open issues about this but I guess the current behaviour is due to this issue: #10991

> Ctrl/Alt + Q/W/E/R don't change sample set anymore

I didn't find a specific discussion about this but I think it is mentioned in the "hotkeys" section of this discussion: #20706

> you can't easily manipulate objects (slider anchors, primarily) off the grid that aren't truly offscreen

I guess it could be addressed by #13189

> Can't manipulate volumes mid-object

#9940

> You can't adjust distance snap or beat snap divisor with scroll wheel

I guess the defaults for this is different from stable.
For distance snap: Ctrl + Alt + Scroll wheel
For beat snap divisor: Ctrl + Shift + Scroll wheel

I think you can change the hotkeys in the input settings.

> You can't set tick rate to 0.5 in the editor

#28519

> I don't see a reason to have the select timeline to drag when that makes the hitbox for selecting multiple objects on the timeline smaller

Maybe #23784

> In stable, it is possible to click the time code in the bottom left to copy the time.

Well, you can copy the time pressing Ctrl + C with no selected objects, but I don't think there is a way to paste it at the moment.

> No video support in the editor

Strangely enough, I didn't find any open issue for this. Maybe I didn't look hard enough. It actually was on #988 on the design section, but that issue got closed and I guess no one created a separate issue for the design section of the editor or just video support.

**Reply by RitaSummers** [reply]
Thank you! I updated the discussion

**Comment by felipemarins** [comment]
> In stable, it is possible to click the time code in the bottom left to copy the time.

With #28511 being merged now, I guess you can remove that line.

Edit:
> Objects don't inherit the volume and sample set of the last objects.

I guess this is addressed by #28728 that was merged a few hours ago. I don't know about the sample set part, but the PR text says something about a "auto bank", so maybe you have to select it in order to inherit the sample set of the previous object. I will test later.
